text/0000-from-lossy.md

```diff
@@ -0,0 +1,416 @@
+/// (as in integer types) or variable (as in floating point types) but should be
+/// signficantly smaller than the magnitude of x (`u << |x|`) and should be
+/// consistent with typical values for the result type.
+pub trait TryFromLossy<S> {
+    type Error;
+    fn try_from_lossy(x: S) -> Result<Self, Self::Error>;
+}
+```
+
+As with `TryFrom`, we propose to add a default implementation once
+specialisation is available:
+```rust
+impl<S, T> TryFromLossy<S> for T where FromLossy<S>: T {
+    default type Error = !;
+    default fn try_from_lossy(value: S) -> Result<Self, Self::Error> {
+        Ok(T::from_lossy(value))
+    }
+}
+```
+
+Add implementations from all of:
+
+- `f32, f64`
+
+to all of:
+
+- `u8, u16, u32, u64, u128`
+- `i8, i16, i32, i64, i128`
+
+(Note: `f32` → `u128` is infallible (but still lossy) *if* the input is not
+negative, infinite or an NaN. So even though the output type has large enough
+range, this conversion trait is still applicable.)
+
+The implementations should fail on NaN, Inf, and values whose integer
+approximation is not representable. The integer approximation should be the
+value rounded towards zero. E.g.:
+
+- 1.6f32 → u32: 1
+- -1f32 → u32: error
+- -0.2f32 → u32: 0
+- 100_000f32 → u16: error
+
+## Other conversions
+
+### `WrappingFrom` trait
+
+There are several types of integer conversion which do not preserve the
+semantic value:
+
+-   sign transmutation (i.e. reinterpretation as another integer type of the
+    same size but with different signed-ness), e.g. `128u8 → -128i8`
+-   unsigned truncation (simply dropping excess high bits), e.g. `260u16 → 4u8`
+-   signed truncation (dropping excess high bits and sign-transmuting the
+    remainder), e.g. `-260i16 → -4i8`
+
+Note that there is less incentive against usage of `as` in these cases since the
+conversions do not preserve "value", although alternatives still have some use
+(e.g. to clarify that a conversion is a truncation).
+
+We can add `std::convert::WrappingFrom`:
+
+```rust
+/// A trait for wrapping conversions; these may truncate (drop high bits) and
+/// sign-convert (re-interpret meaning of high bits).
+/// 
+/// Note that `TryFrom` and `WrappingFrom` may both be implemented for the same
+/// source and target types; in this case they should have equivalent result
+/// where `TryFrom` succeeds.
+pub trait WrappingFrom {
+    fn wrapping_from(x: T) -> Self;
+}
+```
+
+Add implementations for all integer conversions which are *not implemented* by
+`From`. These conversions should truncate or zero-extend to the size of the
+result type, then reinterpret (transmute) the bits to the result type.
+
+Potentially this could also re-implement all `From` conversions on integer
+types. This could be useful for generic code. Unfortunately since Rust has no
+`Integer` trait there is no simple way to re-implement `From` *only* where the
+source and target types are integer types.
```

Comment:
I believe it should have _all_ pairs for the integer types, because a wrapping conversion from u8 to u16 has perfectly-well defined semantics.  (Though I'd also like a clippy lint suggesting `from` instead where possible.)  And I think it could be useful for machine-generated code as mentioned in https://github.com/rust-lang/rfcs/pull/2438#issuecomment-403255258.